### DISCUSSION TITLE
Non stop restarts

### DISCUSSION THREAD

**Discussion by GadiTunes1** [vesoft-inc/nebula#5706] [Q&A]
I watched your the "LlamaIndex Workshop: Building RAG with Knowledge Graph" last week.
It was very interesting, and I've been trying to run the Colab notebook that was shared but I get the following error.

```
**WARNING:nebula3.logger:Connect [127.0.0.1:9669](http://127.0.0.1:9669/) failed: socket error connecting to host 127.0.0.1, port 9669** (('127.0.0.1', 9669)): ConnectionRefusedError(111, 'Connection refused')
The ngql extension is already loaded. To reload it, use:
  %reload_ext ngql
Connect [127.0.0.1:9669](http://127.0.0.1:9669/) failed: socket error connecting to host 127.0.0.1, port 9669 (('127.0.0.1', 9669)): ConnectionRefusedError(111, 'Connection refused')
Connect [127.0.0.1:9669](http://127.0.0.1:9669/) failed: socket error connecting to host 127.0.0.1, port 9669 (('127.0.0.1', 9669)): ConnectionRefusedError(111, 'Connection refused')
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-21-d71ba808a1c0>](https://localhost:8080/#) in <cell line: 4>()
      2 os.environ['GRAPHD_HOST'] = "localhost"
      3 connection_string = f"--address {os.environ['GRAPHD_HOST']} --port 9669 --user root --password {os.environ['NEBULA_PASSWORD']}"
----> 4 get_ipython().run_line_magic('ngql', '{connection_string}')
      5 get_ipython().run_line_magic('ngql', 'USE demo_basketballplayer')

4 frames
<decorator-gen-118> in ngql(self, line, cell, local_ns)

[/usr/local/lib/python3.10/dist-packages/nebula3/gclient/net/ConnectionPool.py](https://localhost:8080/#) in init(self, addresses, configs, ssl_conf)
     73         ok_num = self.get_ok_servers_num()
     74         if ok_num < len(self._addresses):
---> 75             raise RuntimeError(
     76                 'The services status exception: {}'.format(self._get_services_status())
     77             )

RuntimeError: The services status exception: [services: ('127.0.0.1', 9669), status: BAD]

I am using the direct OpenAI solution, not via Azure.
I changed the openrc file to:
export OPENAI_API_KEY="sk-my_real_key"
export OPENAI_API_BASE="https://xxx.openai.azure.com/"
export OPENAI_EMBEDDING_ENGINE="xxx"
export OPENAI_ENGINE="xxx"
export GRAPHD_HOST="127.0.0.1"
export GRAPHD_PORT="9669"
export NEBULA_USER="root"
export NEBULA_PASSWORD="nebula"
export NEBULA_ADDRESS=$GRAPHD_HOST:$GRAPHD_PORT

Where OPENAI_API_KEY contains my real key, all the other OPENAI_ env variables are as above since I use the direct OpenAI solution and not Azure, and I changed the GRAPHD_HOST to "127.0.0.1" (as I assumed 1.2.3.4 is a placeholder). I am using a windows machine.

When I ran "curl -fsSL [nebula-up.siwei.io/install.sh](http://nebula-up.siwei.io/install.sh) | bash" from wsl, I got the following (this is the end of it, the upper part was fine):
[+] Running 1/1 ✔ Container nebula-graph-studio-370-web-1 Started 0.8s ✔️ install_nebula_graph_studio Finished ℹ️ Preparing NebulaGraph Console Script... ℹ️ Pulling nebula-console docker image ℹ️ Created console.sh 😁: ✔️ install_nebula_graph_console Finished ℹ️ **Waiting for all nebula-graph containers to be healthy... ℹ️ Nebula-Graph Containers Healthcheck Attempt: 1 Failed, Retrying in 4 Seconds... ℹ️ Nebula-Graph Containers Healthcheck Attempt: 2 Failed, Retrying in 8 Seconds... ℹ️ Nebula-Graph Containers Healthcheck Attempt: 3 Failed, Retrying in 16 Seconds... ℹ️ Nebula-Graph Containers Healthcheck Attempt: 4 Failed, Retrying in 32 Seconds... ℹ️ Nebula-Graph Containers Healthcheck Attempt: 5 Failed, Retrying in 64 Seconds... ⚠️ Failed to waiting for all containers to be healthy, check docker ps for details. ✔️ waiting_for_nebula_graph_up Finished** ┌────────────────────────────────────────┐ │ 🌌 Nebula-Graph Playground is Up now! │ ├────────────────────────────────────────┤ │ │ │ 🎉 Congrats! Your Nebula is Up now! │ │ $ cd ~/.nebula-up │ │ │ │ 🌏 You can access it from browser: │ │ http://127.0.0.1:7001/ │ │ http://<other_interface>:7001 │ │ │ │ 🔥 Or access via Nebula Console: │ │ $ ~/.nebula-up/console.sh │ │ │ │ To remove the playground: │ │ $ ~/.nebula-up/uninstall.sh │ │ │ │ 🚀 Have Fun! │ │ │
```

When checking the containers, I see that they keep restarting:
```
**Restarting (139) 11 seconds ago**                                                                                                          nebula-docker-compose-metad1-1
d30a1b4d2d20   vesoft/nebula-metad:v3.6.0          "/usr/local/nebula/b…"   5 minutes ago   **Restarting (139) 14 seconds ago**                                                                                                          nebula-docker-compose-metad0-1
46b679c645f0   vesoft/nebula-metad:v3.6.0          "/usr/local/nebula/b…"   5 minutes ago   **Restarting (139) 13 seconds ago**                                                                                                          nebula-docker-compose-metad2-1
```

I am using a windows machine and that's where my Docker Desktop runs.

I hope you can help me resolve this.
Thanks,
Gadi

**Comment by wey-gu** [comment]
Dear @GadiTunes1 ,

Wow, Thanks for trying it!

Could you go to ~/.nebula-up in wsl to see the nebula-docker-compose folder, and check logs under it, we could focus on metad's log here.

Alternatively, for Windows desktops, you could also try with NebulaGraph docker extension instead, nebula-up was more for the server side, I didn't quite test on WSL/WSL2 :-P. (be sure to do uninstall of the nebula-up before doing this)

- https://www.docker.com/blog/distributed-cloud-native-graph-database-nebulagraph-docker-extension/
- https://docs.nebula-graph.io/3.6.0/2.quick-start/1.quick-start-workflow/

Welcome to the NebulaGraph community!

BR//Wey

**Reply by GadiTunes1** [reply]
Thank you.
In WSL, I went to the folder ~/.nebula-up/nebula-docker-compose/logs/meta0 (I hope this is what you meant). There are many logs there. 
I didn't get into all of the ones that have a timestamp in their name.
Here is the content of nebula-metad.ERROR and FATAL and WARNING:
Log file created at: 2023/09/05 12:07:31
Running on machine: d8102a87d00f
Running duration (h:mm:ss): 0:00:00
Log line format: [IWEF]yyyymmdd hh:mm:ss.uuuuuu threadid file:line] msg
F20230905 12:07:31.248116    74 RocksEngine.cpp:82] Check failed: status.ok() IO error: While lock file: /data/meta/nebula/0/data/LOCK: Resource temporarily unavailable

Here is the content of nebula-metad.INFO:
Log file created at: 2023/09/05 12:10:33
Running on machine: d8102a87d00f
Running duration (h:mm:ss): 0:00:00
Log line format: [IWEF]yyyymmdd hh:mm:ss.uuuuuu threadid file:line] msg
I20230905 12:10:33.283515     1 MetaDaemon.cpp:137] localhost = "metad0":9559
I20230905 12:10:33.287107     1 NebulaStore.cpp:48] Start the raft service...
I20230905 12:10:33.287391     1 NebulaSnapshotManager.cpp:25] Send snapshot is rate limited to 10485760 for each part by default
I20230905 12:10:33.442214     1 RaftexService.cpp:46] Start raft service on 9560
I20230905 12:10:33.442360     1 NebulaStore.cpp:82] Scan the local path, and init the spaces_
I20230905 12:10:33.442427     1 NebulaStore.cpp:90] Scan path "/data/meta/nebula/0"
I20230905 12:10:33.442443     1 NebulaStore.cpp:292] Init data from partManager for "metad0":9559
I20230905 12:10:33.442479     1 NebulaStore.cpp:417] Create data space 0
F20230905 12:10:33.444042    74 RocksEngine.cpp:82] Check failed: status.ok() IO error: While lock file: /data/meta/nebula/0/data/LOCK: Resource temporarily unavailable

Similar errors appear in meta1. Should I check anything else?

Thanks,
Gadi

**Comment by GadiTunes1** [comment]
I followed the instructions you sent in the links regarding NebulaGraph docker extension, and over there things seemed fine.
Still, when I run the notebook that was shown in the lecture, I still get:
WARNING:nebula3.logger:Connect 127.0.0.1:9669 failed: socket error connecting to host 127.0.0.1, port 9669 (('127.0.0.1', 9669)): ConnectionRefusedError(111, 'Connection refused')
The ngql extension is already loaded. To reload it, use:
  %reload_ext ngql
Connect 127.0.0.1:9669 failed: socket error connecting to host 127.0.0.1, port 9669 (('127.0.0.1', 9669)): ConnectionRefusedError(111, 'Connection refused')
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-10-d71ba808a1c0>](https://localhost:8080/#) in <cell line: 4>()
      2 os.environ['GRAPHD_HOST'] = "localhost"
      3 connection_string = f"--address {os.environ['GRAPHD_HOST']} --port 9669 --user root --password {os.environ['NEBULA_PASSWORD']}"
----> 4 get_ipython().run_line_magic('ngql', '{connection_string}')
      5 get_ipython().run_line_magic('ngql', 'USE demo_basketballplayer')

4 frames
<decorator-gen-118> in ngql(self, line, cell, local_ns)

[/usr/local/lib/python3.10/dist-packages/nebula3/gclient/net/ConnectionPool.py](https://localhost:8080/#) in init(self, addresses, configs, ssl_conf)
     73         ok_num = self.get_ok_servers_num()
     74         if ok_num < len(self._addresses):
---> 75             raise RuntimeError(
     76                 'The services status exception: {}'.format(self._get_services_status())
     77             )

RuntimeError: The services status exception: [services: ('127.0.0.1', 9669), status: BAD]

**Reply by wey-gu** [reply]
Could you please check the status of the containers of the extension in the Docker desktop?

To see them, you need to configure the docker desktop to do so by referring to https://docs.docker.com/desktop/extensions-sdk/dev/test-debug/#show-the-extension-containers.

<img width="2032" alt="Screenshot 2023-09-06 at 09 45 16" src="https://github.com/vesoft-inc/nebula/assets/1651790/c151bf22-e825-4e8d-b8bb-5ae96b926078">


GraphD is listening and exposing 9669 to the host, I am not sure if Windows docker desktop will forward this properly, too. To bypass this windows host --> docker desktop network issue, maybe running the Jupyter inside the wsl will be better?

**Reply by GadiTunes1** [reply]
Thank you. The containers in the UI seem fine:
![image](https://github.com/vesoft-inc/nebula/assets/131252053/293a3ce0-f061-4734-bd97-8d1bf07e7a09)

I trued running the notebook through WSL. I get a new error:
Connection Pool Created
INFO:nebula3.logger:Get connection to ('127.0.0.1', 9669)
Get connection to ('127.0.0.1', 9669)
[ERROR]:
 'IPythonNGQL' object has no attribute '_decode_value'
INFO:nebula3.logger:Get connection to ('127.0.0.1', 9669)
Get connection to ('127.0.0.1', 9669)

What seems to be the issue?

BTW, I ran the notebook so far on Colab since I have GPU resources there.
Now that I run Jupyter from WSL I will not have a GPU. Does this notebook require one?
If so, how can I use Colab for it?

Thanks,
Gadi

**Reply by wey-gu** [reply]
All calls are OpenAI or Azure OpenAI(i used the latter) based thus no local models(that require GPU) are needed. I only put to colab for its sharing feature.

Another thing you could do is to go from the AI Suite in the extension, click the ai suite tab and click the Install NX Mode, after a few minutes, you should be able to click jupyter NB networkX to access a jupyter notebook env inside the docker network.

<img width="1401" alt="Screenshot 2023-09-07 at 18 29 07" src="https://github.com/vesoft-inc/nebula/assets/1651790/6804560c-003f-4a28-9894-714ad43d8981">

**Comment by GadiTunes1** [comment]
Thank you.
Before trying the AI Suite in the extension in the extension I see there an error about _decode_value. What is the reason?
<img width="778" alt="image" src="https://github.com/vesoft-inc/nebula/assets/131252053/19212b7f-ae5d-4909-880f-17ae8b536424">

And when I try the AI Suite in the extension and click on the Jupyter NB NetworkX button after installing NX mode, I get the following error in the browser:
<img width="643" alt="image" src="https://github.com/vesoft-inc/nebula/assets/131252053/d1c4c40d-6787-48da-ac97-7233832f1838">
What am I supposed to do there?

Thank you,
Gadi
